### ISSUE TITLE
Bug: [no-floating-promises] .catch (with no call expression/parentheses) does not fail the rule

### ISSUE THREAD

**Issue by NotWoods** [typescript-eslint/typescript-eslint#6663]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MYewdgzgLgBADgJxAWwJYQKYwLwwApJqYB0CGEIANgG4YAUAlANwCwAUIiuhu50RsWABDKMAAWQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const promise = Promise.resolve();
promise.catch
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-floating-promises": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

```ts
promise.catch; // should fail
promise.catch(() => {}) // OK
```

Since writing `.catch;` (without parentheses) does not handle rejected promises, it shouldn't pass the linter rule.

### Actual Result

```ts
promise.catch; // OK
promise.catch(() => {}) // OK
```

In our codebase, we found that some engineers would use `promise.catch;` to avoid the linter pointing out that they weren't handling rejected promises. Rather than calling the catch function, they would simply write `.catch` (without parentheses).

### Additional Info

This is caught by no-unused-expressions but it's probably worthwhile to not make this rule dependent on another rule.

**Comment by bradzacher** [comment]
> In our codebase, we found that some engineers would use promise.catch; to avoid the linter pointing out that they weren't handling rejected promises.

I just threw up a little in my mouth. I'm constantly surprised at the lengths some engineers go to to avoid satisfying the linter.

---

This is an interesting thing because **technically** the expression isn't a floating promise - the return type of the expression is a function.
That's why the rule doesn't report on it - the rule inspects the return type of the expression statement and it sees a function type and stops.

So the question is whether or not the rule should specifically consider the user to be a "bad actor" look deeper into the expression to determine if they're playing funny buggers.

I don't think so, personally, because there are numerous ways that a user could escape the rule as a bad actor

```ts
promise.catch
promise.then
promise.constructor
promise.toString

promise.toString()
promise as unknown
promise.catch.length
promise.catch.name
promise.catch.apply
promise.catch.bind
```

Additionally every check we do does increase the cost of the rule because types are computed lazily.
We can't just ignore the `.catch` member because it might be a custom property that's not a promise function like `{ catch: Promise<void>}`, so we'd compute types twice for the expression.

My advice would be to use the `no-unused-expressions` rule in your codebase. It's a good rule in general to help catch coding errors, so it should be on regardless, and it has the added bonus of catching this case and many of the other potential "bad actor" cases.

---

Though TBH this is definitely a huge engineering culture problem and probably warrants a frank discussion with the engineering team.
This code is obviously bad - nobody would argue with it. That fact itself is a red flag to begin with. But to me the bigger issue is that it has also gotten through a code review - which implies either that reviewers are okay with bad code or are stamping without actually reviewing the code.

The engineering team should be striving for well written, bug-free code!

---

If you guys are okay with floating promises, consider turning on the [`ignoreVoid`](https://typescript-eslint.io/rules/no-floating-promises#ignorevoid) option to allow your team to clearly mark promises that they want to be floating.

**Comment by NotWoods** [comment]
Thanks for the though response. As soon as I caught this I turned on `no-unused-expressions`, and it was also helpful in catching a couple copy-paste errors that had slipped in (I was guilty of one, oops). It's something I plan to bring up with our team. I'll be sure to experiment with `ignoreVoid` as well.
